app/controllers/api/pxe_servers_controller.rb

```diff
@@ -1,5 +1,91 @@
 module Api
   class PxeServersController < BaseController
+    INVALID_PXE_SERVER_ATTRS = %w[id href].freeze # Cannot update or create these
+    REQUIRED_PXE_SERVER_ATTRS = %w[name uri].freeze
+    REQUIRED_PXE_MENU_ATTRS = %w[file_name].freeze
+    REQUIRED_PXE_AUTH_ATTRS = %w[userid password].freeze
+
     include Subcollections::PxeImages
+    include Subcollections::PxeMenus
+
+    def create_resource(_type, _id, data = {})
+      PxeServer.transaction do
+        authentication = data.delete('authentication')
+        menus = data.delete('pxe_menus')
+
+        validate_pxe_server_data(data)
+        validate_pxe_create_data(data, REQUIRED_PXE_SERVER_ATTRS, 'pxe server')
+
+        server = collection_class(:pxe_servers).new(data)
+        # generates uir_prefix which is ched if server needs authentication or not
+        server.verify_uri_prefix_before_save
+
+        # assign authentication if needed
+        if server.requires_credentials? && server.missing_credentials?
+          validate_pxe_create_data(authentication || {}, REQUIRED_PXE_AUTH_ATTRS, 'pxe server authentication')
+          server.update_authentication(prepare_credentials(authentication), :save => true)
+        end
+
+        # assign pxe menus if needed
+        server.pxe_menus = create_pxe_menus(menus) if menus
+
+        if server.invalid?
+          raise BadRequestError, "Failed to add a pxe server - #{server.errors.full_messages.join(', ')}"
+        end
+
+        server.save
+        server
+      end
+    end
+
+    def delete_resource(_type, id = nil, data = nil)
+      raise BadRequestError, "Must specify an id for deleting a pxe server" unless id
+
+      super
+    end
+
+    def edit_resource(type, id, data)
+      PxeServer.transaction do
+        server = resource_search(id, type, collection_class(:pxe_servers))
+        menus = data.delete('pxe_menus')
+        authentication = data.delete('authentication')
+        if menus
+          server.pxe_menus.clear
+          data['pxe_menus'] = create_pxe_menus(menus)
+        end
+        server.update_authentication(prepare_credentials(authentication), :save => true) if authentication && server.requires_credentials?
+        server.update!(data)
+        server
+      end
+    end
+
+    private
+
+    def prepare_credentials(authentication)
+      creds = {:default => {}}
+      creds[:default][:userid] = authentication['userid'] if authentication['userid']
+      creds[:default][:password] = authentication['password'] if authentication['password']
+      creds
+    end
+
+    def create_pxe_menus(menus)
+      menus.each do |menu|
+        validate_pxe_create_data(menu, REQUIRED_PXE_MENU_ATTRS, 'pxe menu')
+      end
+      menus.map do |menu|
```

Comment:
this should work also

```ruby
collection_class(:pxe_menus).create(menu)
```

Reply:
in this case, what about to put line ^ on places where `create_pxe_menus`  is called and put validation just before it ?

```ruby
if menus
  server.pxe_menus.clear
  menus.each{|menu| validate_data_for(collection_class(:pxe_menus), menu) }
  data['pxe_menus'] = collection_class(:pxe_menus).create(menu)
end
```